### PR TITLE
Add aggro behavior to monsters

### PR DESCRIPTION
**Pull request by OrenjiAkira**
They now chase your ass down and kill you mercilessly.
This lags if there are lots of monsters chasing you down a narrow one-way path. The workaround for now is limiting their 'vision', and also them killing you before they can group significantly around you.

EDIT: Oh, right. This uses a different distance calculation than we currently to check when a monster is within range. Might need some discussing on whether we want to walk/attack in diagonals.

### CODE REVIEW

File: game/domain/behaviors/aggro.lua [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      found = true
+      goal = current
+      break
+    end
+
+    -- look at neighbors
+    for n = 1, 4 do
+      local i, j = unpack(current)
+      local ti, tj = unpack(goal)
+      local di, dj = unpack(DIR[DIR[n]])
```
Comment (Kazuo256) [inline game/domain/behaviors/aggro.lua]: If you prefer, you can iterate directly over `DIR` instead:
```lua
for _,dir in ipairs(DIR) do
  local di, dj = DIR[dir]
end
```

**Comment by OrenjiAkira** [comment]
WIP or no WIP, I need feedback on those problems I mentioned.

**Comment by Kazuo256** [comment]
I think limiting the vision should make a good difference. If it doesn't, we can either use a relaxed A-star or retort to "dumber" algorithms (as in, walk straight in the target's direction).

The relaxed A-star is very simple, just multiply the weight of the distance heuristic by 2. It stops finding the best paths, but is still better than "dumb" and considerably faster.

As for the walking and attacking diagonally, I'm not 100% sure yet but the more I think about it the more it seems to work better overall.

**Comment by OrenjiAkira** [comment]
That cannot be that easy, I swear to god...

**Comment by OrenjiAkira** [comment]
Ok, I saw literally no difference in multiplying the dist heuristic by two

**Comment by Kazuo256** [comment]
[Like this?](https://en.wikipedia.org/wiki/A*_search_algorithm#Bounded_relaxation)

**Comment by Kazuo256** [comment]
You saw no difference in efficiency?

**Comment by OrenjiAkira** [comment]
I will test it more properly to see, but from playing I still felt around the same ammount of lag

**Comment by OrenjiAkira** [comment]
according to my data, the tests yielded similar lag and time usage.
i'm using code nearly identical to this: https://www.redblobgames.com/pathfinding/a-star/implementation.html#python-astar

**Comment by Kazuo256** [comment]
> multiplying the heuristic would just make priorities have bigger values

It changes how tiles are prioritized in the path, by making tiles "closer" to the target "better", even if they are further away from the starting point.

But if ain't improving speed, then that's that. As far as I can see though, the current implementation still allows actors to "see" the whole map, right?

**Comment by OrenjiAkira** [comment]
do you mean "seeing" by:

1. Seeing a player within a certain distance and then choose to chase it (thus running the algorythm)

or

2. When running the algorythm, not looking at anywhere too far away?

**Comment by Kazuo256** [comment]
> 2. When running the algorythm, not looking at anywhere too far away?

This.

**Comment by OrenjiAkira** [comment]
Then, yes, it does look at the entire map for a possible solution until it is clear that no solution is available.

How would you suggest going around that?

**Comment by OrenjiAkira** [comment]
Should be enough for a merge, but review properly first

**Pull request merged**